tests/testthat/test-topological_sort.R

```diff
@@ -0,0 +1,80 @@
+test_that("topological sort throws error if circular relationship", {
+
+  expect_error(topological_sort(
+    list(n1 = "n3", n2 = "n2", n3 = c(""))
+  ))
+
+  expect_error(topological_sort(
+    list(n1 = "n2", n2 = "n1")
+  ))
+
+  expect_error(topological_sort(
+    list(n1 = "n2", n2 = "n3", n3 = c("n4", "n1"), n4 = c())
+  ))
+
+})
+
+test_that("topological sort correctly sorts graph with no branching", {
+  sorted_deps <- topological_sort(
+    list(
+      n1 = "n2",
+      n2 = "n4",
+      n3 = "n5",
+      n4 = "n3",
+      n5 = c()
+    )
+  )
+  expect_equal(sorted_deps, c("n5", "n3", "n4", "n2", "n1"))
+})
+
+test_that("topological sort correctly sorts disconnected graph", {
+  sorted_deps <- topological_sort(
+    list(
+      n1 = "n2",
+      n2 = "n4",
+      n3 = "n5",
+      n4 = c(),
+      n5 = c()
+    )
+  )
+
+  expect_true(which(sorted_deps == "n2") < which(sorted_deps == "n1"))
+  expect_true(which(sorted_deps == "n4") < which(sorted_deps == "n2"))
+  expect_true(which(sorted_deps == "n5") < which(sorted_deps == "n3"))
+  expect_equal(length(sorted_deps), 5)
+  expect_equal(unique(sorted_deps), sorted_deps)
+})
+
+test_that("topological sort works with graph with branching", {
+  sorted_deps <- topological_sort(
+    list(
+      n1 = "n3",
+      n2 = c("n1", "n4"),
+      n3 = c(),
+      n4 = "n3"
+    )
+  )
+
+  expect_equal(length(sorted_deps), 4)
+  expect_equal(unique(sorted_deps), sorted_deps)
+  expect_equal(sorted_deps[1], "n3")
+  expect_equal(sorted_deps[4], "n2")
+  expect_equal(sort(sorted_deps[2:3]), c("n1", "n4")) # ordering of these does not matter
+})
+
+
+test_that("topological sort works when parent occurs twice in graph", {
```

Comment:
This was the specific example causing tern to fail

Reply:
good